[WFCORE-4361] Add evidence decoders in the Elytron subsystem

fjuma opened the pull request:
https://issues.jboss.org/browse/WFCORE-4361
https://issues.jboss.org/browse/EAP7-1173

This also contains the Elytron subsystem version bump included in #3777

To pass CI, this will require an Elytron upgrade but this is ready for the subsystem changes to be reviewed.

Analysis document: https://github.com/wildfly/wildfly-proposals/pull/203
Capabilities PR: https://github.com/wildfly/wildfly-capabilities/pull/34
WildFly PR: https://github.com/wildfly/wildfly/pull/12352
Depends on https://github.com/wildfly-security/wildfly-elytron/pull/1294

jmesnil commented:
@fjuma I have merged #3777 (electron subsystem bump) and #3821 (elytron component upgrades), could you please remove the subsystem bump commits from this PR please?

darranl commented:
@jmesnil This is a GitHub bug not an issue with this PR, I will have a look at the current PRs.

OndrejKotek commented on elytron/src/main/java/org/wildfly/extension/elytron/EvidenceDecoderDefinitions.java:
```diff
@@ -0,0 +1,147 @@
+            .setRestartAllServices()
+            .build();
+
+    static final SimpleAttributeDefinition SEGMENT = new SimpleAttributeDefinitionBuilder(ElytronDescriptionConstants.SEGMENT, ModelType.INT, true)
```
It could be checked whether the value is non-negative (>=0).

fjuma replied:
Thanks @OndrejKotek, fixed.

darranl commented:
@fjuma you also have a conflict but may be worth waiting until after the 1.10.0.CR2 tag as we have another imminent merge likely to conflict with this PR.

fjuma commented:
I've rebased this one and fixed the conflict.

darranl commented:
@fjuma With this capability I am just wondering should there be a custom-evidence-decoder resource like we have for the other capabilities?

fjuma commented:
@darranl This is certainly something that could be added. Currently, the only type of evidence that we have that has a principal associated with it is is ```X509PeerCertificateChainEvidence```. The evidence decoders added in this PR now cover the identifying information that can be extracted from an X.509 certificate (the subject name and subject alternative names). However, custom evidence decoders could be useful for custom evidence types. I can add this in.

fjuma commented:
@darranl I've updated this PR to add support for a ```custom-evidence-decoder```. I've also updated the analysis document and documentation PR accordingly as well. 

@OndrejKotek ^ FYI. This is a small change.

darranl commented:
Follow up PR available here https://github.com/wildfly/wildfly-core/pull/3845 addressing merge conflict.
